File: engine/src/scripting/managed/Components/Transform.cs
//// Transform.cs /////////////////////////////////////////////////////////////
//
//  zzzzz       zzz  zzzzzzzzzzzzz    zzzz      zzzz       zzzzzz  zzzzz
//  zzzzzzz     zzz  zzzz                    zzzz       zzzz           zzzz
//  zzz   zzz   zzz  zzzzzzzzzzzzz         zzzz        zzzz             zzz
//  zzz    zzz  zzz  z                  zzzz  zzzz      zzzz           zzzz
//  zzz         zzz  zzzzzzzzzzzzz    zzzz       zzz      zzzzzzz  zzzzz
//
//  Author:      Guillaume HEIN
//  Date:        03/05/2025
//  Description: Source file for the Transform component in C#.
//
///////////////////////////////////////////////////////////////////////////////

using System.Numerics;
using System.Runtime.InteropServices;

namespace Nexo.Components
{

    [StructLayout(LayoutKind.Sequential)]
    public struct Transform
    {
        public Vector3 pos;
        public Vector3 size;
        public Quaternion quat;
        
        public Matrix4x4 worldMatrix;
        public Matrix4x4 localMatrix;

        public Vector3 localCenter;

        // Note: children is a C++ vector class, it cannot be directly represented in C#.
        // TODO: Find a way to handle vectors and lists between C++ and C#.
    }

}

File: engine/src/components/Transform.cpp
//// Transform.cpp ///////////////////////////////////////////////////////////////
//
//  zzzzz      zzz   zzzzzzzzzzzzz   zzzz      zzzz      zzzzz  zzzzz
//  zzz zzz    zzz   zzzz              zzzz  zzzz      zzzz        zzzz
//  zzz   zzz  zzz   zzzzzzzzz            zzzz        zzzz          zzzz
//  zzz    zzz zzz   zzzz              zzzz  zzzz      zzzz        zzzz
//  zzz      zzzzz   zzzzzzzzzzzzz   zzzz      zzzz      zzzzz  zzzzz
//
//  Author:      Mehdy MORVAN
//  Date:        11/07/2025
//  Description: Source file for the transform component
//
///////////////////////////////////////////////////////////////////////////////

#include "Transform.hpp"

#include <algorithm>

namespace nexo::components {
    void TransformComponent::restore(const TransformComponent::Memento &memento)
    {
        pos = memento.position;
        quat = memento.rotation;
        size = memento.scale;
        localMatrix = memento.localMatrix;
        localCenter = memento.localCenter;
        children = memento.children;
    }

    [[nodiscard]] TransformComponent::Memento TransformComponent::save() const
    {
        return {pos, quat, size, localMatrix, localCenter, children};
    }

    void TransformComponent::addChild(const ecs::Entity childEntity)
    {
        if (std::ranges::find(children, childEntity) != children.end())
            return;
        children.push_back(childEntity);
    }

    void TransformComponent::removeChild(const ecs::Entity childEntity)
    {
        children.erase(std::ranges::remove(children, childEntity).begin(), children.end());
    }
}
